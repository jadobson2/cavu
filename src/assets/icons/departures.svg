<svg width="51" height="48" viewBox="0 0 51 48" fill="none" xmlns="http://www.w3.org/2000/svg">
<line x1="4" y1="42" x2="48" y2="42" stroke="black" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M46.5269 11.3757L45.7671 13.2257L46.5269 11.3757ZM47.3825 16.391L46.0523 14.8975L47.3825 16.391ZM40.3332 10.1196L39.5496 8.27957L40.3332 10.1196ZM43.4201 10.0997L42.6603 11.9498L43.4201 10.0997ZM27.7973 14.4164L28.6125 12.5901L27.7973 14.4164ZM30.1956 14.437L30.9792 16.2771L30.1956 14.437ZM7.54262 8.86705L8.32628 10.7071L7.54262 8.86705ZM18.3599 21.65L17.576 19.81L18.3599 21.65ZM18.6415 20.2098L17.2814 21.6761L18.6415 20.2098ZM12.7476 22.3978L11.5388 23.9912L12.7476 22.3978ZM3.16649 21.972L1.70508 23.3374L3.16649 21.972ZM3.50538 20.3693L2.72173 18.5292L3.50538 20.3693ZM11.1486 29.662L10.7797 31.6277L11.1486 29.662ZM9.50985 28.7615L8.04844 30.1269L9.50985 28.7615ZM16.3765 30.6431L16.0076 32.6088L16.3765 30.6431ZM18.1054 30.4547L17.3217 28.6147L18.1054 30.4547ZM44.4402 19.0116L43.11 17.5181L44.4402 19.0116ZM43.3471 19.7047L42.5635 17.8646L43.3471 19.7047ZM16.7454 28.6775L11.5175 27.6963L10.7797 31.6277L16.0076 32.6088L16.7454 28.6775ZM16.5206 24.6076L19.1438 23.4899L17.576 19.81L14.9527 20.9277L16.5206 24.6076ZM12.7216 9.87728L26.9821 16.2427L28.6125 12.5901L14.352 6.22465L12.7216 9.87728ZM30.9792 16.2771L41.1169 11.9597L39.5496 8.27957L29.4119 12.597L30.9792 16.2771ZM42.6603 11.9498L45.7671 13.2257L47.2867 9.5256L44.18 8.24966L42.6603 11.9498ZM46.0523 14.8975L43.11 17.5181L45.7705 20.505L48.7127 17.8845L46.0523 14.8975ZM42.5635 17.8646L17.3217 28.6147L18.889 32.2948L44.1308 21.5448L42.5635 17.8646ZM10.9713 27.3962L4.62791 20.6067L1.70508 23.3374L8.04844 30.1269L10.9713 27.3962ZM20.0015 18.7434C17.509 16.4316 11.2331 11.2193 8.5669 9.01635L6.01908 12.1C8.72689 14.3373 14.8938 19.4616 17.2814 21.6761L20.0015 18.7434ZM8.32628 10.7071L10.3233 9.85664L8.75597 6.17649L6.75897 7.02697L8.32628 10.7071ZM4.28902 22.2094L5.82621 21.5547L4.25891 17.8746L2.72173 18.5292L4.28902 22.2094ZM8.81481 21.9248L11.5388 23.9912L13.9563 20.8044L11.2323 18.738L8.81481 21.9248ZM5.82621 21.5547C6.81664 21.1329 7.95715 21.2741 8.81481 21.9248L11.2323 18.738C9.23112 17.2199 6.56993 16.8904 4.25891 17.8746L5.82621 21.5547ZM45.7671 13.2257C46.4506 13.5064 46.6041 14.406 46.0523 14.8975L48.7127 17.8845C51.4717 15.4271 50.7044 10.9292 47.2867 9.5256L45.7671 13.2257ZM41.1169 11.9597C41.6093 11.75 42.1652 11.7464 42.6603 11.9498L44.18 8.24966C42.6947 7.63965 41.0268 7.65042 39.5496 8.27957L41.1169 11.9597ZM26.9821 16.2427C28.2517 16.8094 29.7 16.8219 30.9792 16.2771L29.4119 12.597C29.1561 12.7059 28.8664 12.7034 28.6125 12.5901L26.9821 16.2427ZM14.352 6.22465C12.5745 5.43122 10.5469 5.41377 8.75597 6.17649L10.3233 9.85664C11.0908 9.52976 11.9598 9.53724 12.7216 9.87728L14.352 6.22465ZM8.5669 9.01635C9.157 9.50392 8.99877 10.4207 8.32628 10.7071L6.75897 7.02697C4.70496 7.90174 4.2714 10.656 6.01908 12.1L8.5669 9.01635ZM19.1438 23.4899C20.9601 22.7161 21.6447 20.2674 20.0015 18.7434L17.2814 21.6761C17.0113 21.4256 16.8664 21.0147 16.9495 20.6043C17.0292 20.2105 17.2928 19.9306 17.576 19.81L19.1438 23.4899ZM14.9527 20.9277C14.6225 21.0684 14.2423 21.0213 13.9563 20.8044L11.5388 23.9912C12.9685 25.0757 14.8697 25.311 16.5206 24.6076L14.9527 20.9277ZM4.62791 20.6067C5.10169 21.1138 4.92752 21.9375 4.28902 22.2094L2.72173 18.5292C0.806239 19.345 0.283738 21.8161 1.70508 23.3374L4.62791 20.6067ZM11.5175 27.6963C11.3081 27.657 11.1167 27.5518 10.9713 27.3962L8.04844 30.1269C8.77573 30.9054 9.73264 31.4312 10.7797 31.6277L11.5175 27.6963ZM16.0076 32.6088C16.978 32.7909 17.9806 32.6817 18.889 32.2948L17.3217 28.6147C17.14 28.692 16.9395 28.7139 16.7454 28.6775L16.0076 32.6088ZM43.11 17.5181C42.9479 17.6625 42.7633 17.7796 42.5635 17.8646L44.1308 21.5448C44.7302 21.2895 45.284 20.9383 45.7705 20.505L43.11 17.5181Z" fill="black"/>
</svg>
