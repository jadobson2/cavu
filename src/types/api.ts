import { Departure } from '.'

export interface GetDeparturesResponse {
  allDepartures: Departure[]
}
