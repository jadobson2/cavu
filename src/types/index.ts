export interface ArrivalAirport {
  name: string;
  cityName: string;
  countryName: string;
  code: string;
}

export interface DepartureAirport {
  name: string;
  cityName: string;
  countryName: string;
  code: string;
}

export interface Airline {
  name: string;
  code: string;
}

export interface DepartureGate {
  name: string;
  number: string;
  action: string;
}

export interface Departure {
  flightDirection: string;
  scheduledDepartureDateTime: string;
  scheduledArrivalDateTime: string;
  estimatedDepartureDateTime: string;
  actualDepartureDateTime: string;
  arrivalAirport: ArrivalAirport;
  departureAirport: DepartureAirport;
  flightNumber: string;
  airline: Airline;
  departureGate: DepartureGate | null;
  arrivalTerminal: string | null;
  departureTerminal: string;
  status: string;
}

export enum DepartureStatusCode {
  BOARDING = 'BOARDING',
  CANCELLED = 'CANCELLED',
  DELAYED = 'DELAYED',
  DEPARTED = 'DEPARTED',
  GATE_CLOSED = 'GATE_CLOSED',
  GATE_CLOSING = 'GATE_CLOSING',
  GATE_OPEN = 'GATE_OPEN'
}
