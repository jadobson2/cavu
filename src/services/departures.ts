import { API_BASE_URL } from '@/constants'
import { Departure } from '@/types'
import { GetDeparturesResponse } from '@/types/api'

export const getDepartures = async (): Promise<Departure[]> => {
  const response = await fetch(`${API_BASE_URL}/departures/Flightdata`)
  const departures: GetDeparturesResponse = await response.json()
  return departures.allDepartures
}
