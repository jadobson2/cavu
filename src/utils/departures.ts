import { DepartureStatusCode } from '@/types'

export const getStatus = (statusText: string): DepartureStatusCode => {
  if (statusText.includes('Boarding')) return DepartureStatusCode.BOARDING
  if (statusText.includes('Cancelled')) return DepartureStatusCode.CANCELLED
  if (statusText.includes('Delayed')) return DepartureStatusCode.DELAYED
  if (statusText.includes('Departed')) return DepartureStatusCode.DEPARTED
  if (statusText.includes('Final Call')) return DepartureStatusCode.GATE_CLOSING
  if (statusText.includes('Go to Gate')) return DepartureStatusCode.GATE_OPEN

  return DepartureStatusCode.GATE_CLOSED
}
