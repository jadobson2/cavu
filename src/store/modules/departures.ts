import { Commit } from 'vuex'
import { getDepartures } from '@/services/departures'
import { Departure } from '@/types'

const state = () => ({
  departures: [] as Departure[],
  error: '',
  loading: false
})

export type DeparturesState = ReturnType<typeof state>

const departures = {
  namespaced: true,
  state,
  mutations: {
    setDepartures (state: DeparturesState, departures: Departure[]) {
      state.departures = departures
    },
    setError (state: DeparturesState, message: string) {
      state.error = message
    },
    setLoading (state: DeparturesState, loading: boolean) {
      state.loading = loading
    }
  },
  actions: {
    async getDepartures ({ commit }: { commit: Commit }) {
      commit('setError', '')
      commit('setLoading', true)

      try {
        const departures = await getDepartures()

        commit('setDepartures', departures)
      } catch {
        commit('setError', 'An error occurred getting departures.')
      } finally {
        commit('setLoading', false)
      }
    },
    updateDeparture (
      { commit, state }: { commit: Commit, state: DeparturesState },
      update: { flightNumber: string, status: string }
    ): boolean {
      const departures = [...state.departures]
      const departure = departures.find((departure) => {
        return departure.flightNumber === update.flightNumber
      })
      if (!departure) return false
      departures[departures.indexOf(departure)] = {
        ...departure,
        status: update.status
      }

      commit('setDepartures', departures)

      return true
    }
  }
}

export default departures
