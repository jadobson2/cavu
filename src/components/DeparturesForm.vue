<template>
  <form class="px-8 py-4 lg:py-8 xl:px-14 rounded-xl bg-white">
    <h2 class="mb-6 text-2xl font-bold">Update Departure</h2>
    <div class="mb-4">
      <label class="form-label" for="update-flight">Flight</label>
      <select v-model="flightNumber" class="form-input" id="update-flight">
        <option :value="null" disabled selected hidden>Select Flight</option>
        <option
          v-for="departure in departures"
          :key="departure.flightNumber"
          :value="departure.flightNumber"
        >
          {{ `
            ${getHours(departure.scheduledDepartureDateTime)}
            -
            ${departure.arrivalAirport.cityName}
          ` }}
        </option>
      </select>
    </div>
    <div class="mb-4 lg:mb-6">
      <label class="form-label" for="update-status">Status</label>
      <select v-model="status" class="form-input" id="update-status">
        <option :value="null" disabled selected hidden>Select Status</option>
        <option
          v-for="option in statusOptions"
          :key="option"
          :value="option"
        >
          {{ option }}
        </option>
      </select>
    </div>
    <div v-if="isStatusOther" class="mb-4 lg:mb-6">
      <label class="form-label" for="update-status-other">Other Status</label>
      <input v-model="statusOther" class="form-input" id="update-status-other"/>
    </div>
    <p
      v-if="success"
      class="mb-4 lg:mb-6 text-green font-bold"
      data-test="success-notice"
    >
      Departure updated successfully.
    </p>
    <p
      v-if="error"
      class="mb-4 lg:mb-6 text-red font-bold"
      data-test="error-notice"
    >
      {{ error }}
    </p>
    <div class="text-right">
      <Button
        type="button"
        data-test="update-button"
        @click="updateDeparture"
      >
        Update Departure
      </Button>
    </div>
  </form>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapActions, mapState } from 'vuex'
import { getHours } from '@/utils/date'
import Button from './Button.vue'

enum DepartureStatusOptions {
  CANCELLED = 'Cancelled',
  DELAYED = 'Delayed',
  DEPARTED = 'Departed',
  DIVERTED = 'Diverted',
  OTHER = 'Other'
}

export default Vue.extend({
  name: 'DeparturesForm',
  components: { Button },
  data () {
    return {
      error: '',
      flightNumber: null as string | null,
      status: null as DepartureStatusOptions | null,
      statusOther: null as string | null,
      statusOptions: Object.values(DepartureStatusOptions),
      success: false
    }
  },
  computed: {
    ...mapState('departures', ['departures']),
    isDisabled (): boolean {
      return !this.flightNumber || !this.status || (this.isStatusOther && this.statusOther === '')
    },
    isStatusOther (): boolean {
      return this.status === DepartureStatusOptions.OTHER
    },
    hasNotice (): boolean {
      return !!this.error || this.success
    }
  },
  watch: {
    flightNumber (flightNumber: string) {
      if (this.hasNotice && flightNumber) this.resetNotices()
    },
    status (status: DepartureStatusOptions) {
      if (this.hasNotice && status) this.resetNotices()
      this.statusOther = null
    }
  },
  methods: {
    ...mapActions('departures', { updateDepartureAction: 'updateDeparture' }),
    getHours,
    async updateDeparture () {
      const updated = await this.updateDepartureAction({
        flightNumber: this.flightNumber,
        status: this.statusOther || this.status
      })

      if (updated) {
        this.success = true
        this.reset()
      } else {
        this.error = 'An error occurred updating the departure.'
      }
    },
    reset () {
      this.flightNumber = null
      this.status = null
      this.statusOther = null
    },
    resetNotices () {
      this.error = ''
      this.success = false
    }
  }
})
</script>

<style lang="scss" scoped>
@import './../assets/styles/shared/form.scss'
</style>
