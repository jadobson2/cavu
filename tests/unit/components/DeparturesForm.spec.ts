import Vuex, { StoreOptions } from 'vuex'
import flushPromises from 'flush-promises'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import DeparturesForm from '@/components/DeparturesForm.vue'
import departuresModule from '@/store/modules/departures'
import { departure } from 'tests/unit/fixtures/departures'
import { RelaxedWrapper } from 'tests/unit/types'

describe('DeparturesForm', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)

  const state = {
    departures: [departure],
    error: '',
    loading: false
  }

  const setup = (testStoreOptions?: StoreOptions<unknown>): RelaxedWrapper => {
    const store = new Vuex.Store({
      modules: {
        departures: {
          ...departuresModule,
          state,
          ...testStoreOptions
        }
      }
    })

    return shallowMount(DeparturesForm, {
      localVue,
      store
    })
  }

  it('renders correctly', () => {
    const wrapper = setup()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('update', async () => {
    const wrapper = setup()
    const flightNumberSelect = wrapper.find('#update-flight')
    const statusSelect = wrapper.find('#update-status')
    await flightNumberSelect.find('option[value="AF1669"]').setSelected()
    await statusSelect.find('option[value="Departed"]').setSelected()
    wrapper.find('[data-test="update-button"]').vm.$emit('click')
    await flushPromises()

    expect(state.departures[0].status).toEqual('Departed')
    expect(wrapper.find('[data-test="success-notice"]').exists()).toBe(true)
    /* eslint-disable @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    expect(flightNumberSelect.element.value).toBe('')
    // @ts-ignore
    expect(statusSelect.element.value).toBe('')
    /* eslint-enable @typescript-eslint/ban-ts-comment */
  })

  test('update failed', async () => {
    const wrapper = setup({
      actions: {
        updateDeparture: jest.fn().mockReturnValue(false)
      }
    })
    await wrapper.find('#update-flight').find('option[value="AF1669"]').setSelected()
    await wrapper.find('#update-status').find('option[value="Departed"]').setSelected()
    wrapper.find('[data-test="update-button"]').vm.$emit('click')
    await flushPromises()

    expect(wrapper.find('[data-test="error-notice"]').text()).toBe(
      'An error occurred updating the departure.'
    )
  })

  test('other status', async () => {
    const wrapper = setup()
    await wrapper.find('#update-flight').find('option[value="AF1669"]').setSelected()
    await wrapper.find('#update-status').find('option[value="Other"]').setSelected()
    await wrapper.find('#update-status-other').setValue('Other status')
    wrapper.find('[data-test="update-button"]').vm.$emit('click')
    await flushPromises()

    expect(state.departures[0].status).toEqual('Other status')
    expect(wrapper.find('#update-status-other').exists()).toBe(false)
    expect(wrapper.vm.statusOther).toBe(null)
  })

  describe('flight number change', () => {
    it('resets success notice if set', async () => {
      const wrapper = setup()
      await wrapper.setData({ success: true })
      await wrapper.find('#update-flight').find('option[value="AF1669"]').setSelected()

      expect(wrapper.find('[data-test="success-notice"]').exists()).toBe(false)
    })

    it('resets error notice if set', async () => {
      const wrapper = setup()
      await wrapper.setData({ error: 'Error message' })
      await wrapper.find('#update-flight').find('option[value="AF1669"]').setSelected()

      expect(wrapper.find('[data-test="success-error"]').exists()).toBe(false)
    })
  })

  describe('status change', () => {
    it('resets status other', async () => {
      const wrapper = setup()
      await wrapper.setData({
        status: 'Other',
        statusOther: 'Other status'
      })
      await wrapper.find('#update-status').find('option[value="Departed"]').setSelected()

      expect(wrapper.find('#update-status-other').exists()).toBe(false)
      expect(wrapper.vm.statusOther).toBe(null)
    })

    it('resets success notice if set', async () => {
      const wrapper = setup()
      await wrapper.setData({ success: true })
      await wrapper.find('#update-status').find('option[value="Departed"]').setSelected()

      expect(wrapper.find('[data-test="success-notice"]').exists()).toBe(false)
    })

    it('resets error notice if set', async () => {
      const wrapper = setup()
      await wrapper.setData({ error: 'Error message' })
      await wrapper.find('#update-status').find('option[value="Departed"]').setSelected()

      expect(wrapper.find('[data-test="success-error"]').exists()).toBe(false)
    })
  })
})
