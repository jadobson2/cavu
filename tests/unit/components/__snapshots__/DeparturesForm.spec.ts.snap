// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeparturesForm renders correctly 1`] = `
<form
  class="px-8 py-4 lg:py-8 xl:px-14 rounded-xl bg-white"
>
  <h2
    class="mb-6 text-2xl font-bold"
  >
    Update Departure
  </h2>
   
  <div
    class="mb-4"
  >
    <label
      class="form-label"
      for="update-flight"
    >
      Flight
    </label>
     
    <select
      class="form-input"
      id="update-flight"
    >
      <option
        disabled="disabled"
        hidden="hidden"
        selected="selected"
        value=""
      >
        Select Flight
      </option>
       
      <option
        value="AF1669"
      >
        
        
          10:30
          -
          Paris
        
      
      </option>
    </select>
  </div>
   
  <div
    class="mb-4 lg:mb-6"
  >
    <label
      class="form-label"
      for="update-status"
    >
      Status
    </label>
     
    <select
      class="form-input"
      id="update-status"
    >
      <option
        disabled="disabled"
        hidden="hidden"
        selected="selected"
        value=""
      >
        Select Status
      </option>
       
      <option
        value="Cancelled"
      >
        
        Cancelled
      
      </option>
      <option
        value="Delayed"
      >
        
        Delayed
      
      </option>
      <option
        value="Departed"
      >
        
        Departed
      
      </option>
      <option
        value="Diverted"
      >
        
        Diverted
      
      </option>
      <option
        value="Other"
      >
        
        Other
      
      </option>
    </select>
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="text-right"
  >
    <button-stub
      data-test="update-button"
      type="button"
    >
      
      Update Departure
    
    </button-stub>
  </div>
</form>
`;
